asserts/snapasserts/validation_sets.go

```diff
@@ -449,3 +449,87 @@ func (v *ValidationSets) CheckInstalledSnaps(snaps []*InstalledSnap) error {
 	}
 	return nil
 }
+
+type PresenceConstraintError struct {
+	SnapName string
+	Presence asserts.Presence
+}
+
+func (e *PresenceConstraintError) Error() string {
+	return fmt.Sprintf("unexpected presence %q for snap %q", e.Presence, e.SnapName)
+}
+
+func (v *ValidationSets) constraintsForSnap(snapRef naming.SnapRef) *snapContraints {
+	if snapRef.ID() != "" {
+		return v.snaps[snapRef.ID()]
+	}
+	// snapID not available, find by snap name
+	for _, cstrs := range v.snaps {
+		if cstrs.name == snapRef.SnapName() {
+			return cstrs
+		}
+	}
+	return nil
+}
+
+// CheckPresenceRequired returns the list of all validation sets that declare
+// presence of the given snap as required and the required revision (or
+// snap.R(0) if no specific revision is required). PresenceConstraintError is
+// returned if presence of the snap is "invalid".
+// The method assumes that validation sets are not in conflict.
+func (v *ValidationSets) CheckPresenceRequired(snapRef naming.SnapRef) ([]string, snap.Revision, error) {
+	cstrs := v.constraintsForSnap(snapRef)
+	if cstrs == nil {
+		return nil, unspecifiedRevision, nil
+	}
+	if cstrs.presence == asserts.PresenceInvalid {
+		return nil, unspecifiedRevision, &PresenceConstraintError{snapRef.SnapName(), cstrs.presence}
+	}
+	if cstrs.presence != asserts.PresenceRequired {
+		return nil, unspecifiedRevision, nil
```

Comment:
afaiu presence has only 3 states, so maybe a comment:
```suggestion
		// optional presence
		return nil, unspecifiedRevision, nil
```